Importing the dependencies

In [ ]:
import numpy as np
import pandas as pd
from  sklearn.model_selection import train_test_split
from sklearn.preprocessing import *                            #preprocessing of model using StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.svm import SVC

Data Collection and Processing

In [ ]:
#loading the csv data to pandas dataframe
heart_data=pd.read_csv('/content/heart.csv')

In [ ]:
#print first five rows of dataset
print(heart_data.head())

   age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
0   63    1   3       145   233    1  ...      0      2.3      0   0     1       1
1   37    1   2       130   250    0  ...      0      3.5      0   0     2       1
2   41    0   1       130   204    0  ...      0      1.4      2   0     2       1
3   56    1   1       120   236    0  ...      0      0.8      2   0     2       1
4   57    0   0       120   354    0  ...      1      0.6      2   0     2       1

[5 rows x 14 columns]


In [ ]:
#print last five rows of dataset
print(heart_data.tail())

     age  sex  cp  trestbps  chol  fbs  ...  exang  oldpeak  slope  ca  thal  target
298   57    0   0       140   241    0  ...      1      0.2      1   0     3       0
299   45    1   3       110   264    0  ...      0      1.2      1   0     3       0
300   68    1   0       144   193    1  ...      0      3.4      1   2     3       0
301   57    1   0       130   131    0  ...      1      1.2      1   1     3       0
302   57    0   1       130   236    0  ...      0      0.0      1   1     2       0

[5 rows x 14 columns]


In [ ]:
#number of rows and columns
heart_data.shape

(303, 14)

In [ ]:
#getting some info about the dataset
heart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


In [ ]:
#checking for missing values
heart_data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [ ]:
#statistical measures about dataset
heart_data.describe()

age         sex          cp  ...          ca        thal      target
count  303.000000  303.000000  303.000000  ...  303.000000  303.000000  303.000000
mean    54.366337    0.683168    0.966997  ...    0.729373    2.313531    0.544554
std      9.082101    0.466011    1.032052  ...    1.022606    0.612277    0.498835
min     29.000000    0.000000    0.000000  ...    0.000000    0.000000    0.000000
25%     47.500000    0.000000    0.000000  ...    0.000000    2.000000    0.000000
50%     55.000000    1.000000    1.000000  ...    0.000000    2.000000    1.000000
75%     61.000000    1.000000    2.000000  ...    1.000000    3.000000    1.000000
max     77.000000    1.000000    3.000000  ...    4.000000    3.000000    1.000000

[8 rows x 14 columns]

In [ ]:
#checking the distribution of target variable
heart_data['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

In [ ]:
#splitting the features and target
X = heart_data.drop(columns='target', axis=1)
Y = heart_data['target']

In [ ]:
X

age  sex  cp  trestbps  chol  ...  exang  oldpeak  slope  ca  thal
0     63    1   3       145   233  ...      0      2.3      0   0     1
1     37    1   2       130   250  ...      0      3.5      0   0     2
2     41    0   1       130   204  ...      0      1.4      2   0     2
3     56    1   1       120   236  ...      0      0.8      2   0     2
4     57    0   0       120   354  ...      1      0.6      2   0     2
..   ...  ...  ..       ...   ...  ...    ...      ...    ...  ..   ...
298   57    0   0       140   241  ...      1      0.2      1   0     3
299   45    1   3       110   264  ...      0      1.2      1   0     3
300   68    1   0       144   193  ...      0      3.4      1   2     3
301   57    1   0       130   131  ...      1      1.2      1   1     3
302   57    0   1       130   236  ...      0      0.0      1   1     2

[303 rows x 13 columns]

In [ ]:
Y

0      1
1      1
2      1
3      1
4      1
      ..
298    0
299    0
300    0
301    0
302    0
Name: target, Length: 303, dtype: int64

Splitting the data into Training & Testing Data

In [ ]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2, stratify=Y,random_state=2)

In [ ]:
#check the training and test data
print(X.shape,X_train.shape,X_test.shape)

(303, 13) (242, 13) (61, 13)


Model Training using Logistic Regression

In [ ]:
lrmodel = LogisticRegression()
#training the LR model
lrmodel.fit(X_train, Y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [ ]:
#preprocessing using standard scaler
rbfmodel=Pipeline([('scaler',StandardScaler()),('svc',SVC(kernel='rbf',gamma='auto'))])
#training the rbf kernel of svm model
rbfmodel.fit(X_train,Y_train)

Pipeline(memory=None,
         steps=[('scaler',
                 StandardScaler(copy=True, with_mean=True, with_std=True)),
                ('svc',
                 SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None,
                     coef0=0.0, decision_function_shape='ovr', degree=3,
                     gamma='auto', kernel='rbf', max_iter=-1, probability=False,
                     random_state=None, shrinking=True, tol=0.001,
                     verbose=False))],
         verbose=False)

In [ ]:
rbftrain=rbfmodel.score(X_train,Y_train)
rbftest=rbfmodel.score(X_test,Y_test)
lrtrain=lrmodel.score(X_train,Y_train)
lrtest=lrmodel.score(X_test,Y_test)

In [ ]:
#printing the models' accuracy
print("RBF Training : ",rbftrain)
print("RBF Testing : ",rbftest)

RBF Training :  0.9132231404958677
RBF Testing :  0.8032786885245902


In [ ]:
print("LR Training : ",lrtrain)
print("LR Testing : ",lrtest)

LR Training :  0.8512396694214877
LR Testing :  0.819672131147541


RBF gives better result over LR Model and we will proceed with RBF kernel SVM Model

Building a Prediction System

In [ ]:
#trying to predict the heart status of new patient
input_data = (52,0,1,125,180,0,0,192,0,1.3,2,0,2)

#change the input to a numpy array
input_data_as_nparray=np.asarray(input_data)

#reshape of the input np array as we predict only one record
input_data_reshape=input_data_as_nparray.reshape(1,-1)

prediction=rbfmodel.predict(input_data_reshape)
print(prediction)

#predicted statements
if(prediction==1):
  print("Patient has heart disease")
else:
  print("Patient is free from heart disease")

[1]
Patient has heart disease


Hence the Prediction System for Heart Disease works efficiently with accuracy greater than 82 %

Designed and Implemented by Rohith Kumar R (Sem 5,PES University,Bengaluru)